src/hooks/use-dropdown-menu.tsx

```diff
@@ -0,0 +1,153 @@
+// Imports
+import React, { useState, useRef, createRef, useEffect } from 'react';
+
+
+// A custom Hook that abstracts away the listeners/controls for dropdown menus
+export default (itemCount: number) => {
+   // Use state
+   const [isOpen, setIsOpen] = useState<boolean>(false);
+   const [returnFocusToButton, setReturnFocusToButton] = useState<boolean>(false);
+   const currentFocusIndex = useRef<number | null>(null);
+   const firstRun = useRef(true);
+   
+   
+   // Create refs
+   const buttonRef = useRef<HTMLButtonElement>(null);
+   const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+   
+   
+   // Create type guard
+   const isKeyboardEvent = (e: React.KeyboardEvent | React.MouseEvent): e is React.KeyboardEvent => (e as React.KeyboardEvent).key !== undefined;
+   
+   
+   // Handle changes in the state of the menu
+   useEffect(() => {
+       // Stop if the ref currently points to a null value
+       if (!itemRefs.current[0].current || !buttonRef.current) {
+           return;
+       }
+
+
+       // Stop if this is the first fire of the Hook, and update the ref
+       if (firstRun.current) {
+           firstRun.current = false;
+           return;
+       }
+
+       
+       // If the menu is currently open focus on the first item in the menu, else return focus to the button if state indicates that is desired behavior
+       if (isOpen) {
+           itemRefs.current[0].current.focus();
+           moveFocus(0);
+       } else if (returnFocusToButton) {
+           buttonRef.current.focus();
+       }
+   }, [isOpen, returnFocusToButton]);
+
+
+   const moveFocus = (itemIndex: number) => {
```

Comment:
This doesn't "Handle changes in the state of the menu" so it needs its own comment.